### PR TITLE
Load balancers v2 documentation

### PR DESCRIPTION
**Pull request by rbibby**

### CODE REVIEW

File: source/network/load-balancers/key-terms.md
Comment (bluemorbo) [inline source/network/load-balancers/key-terms.md]: I think this could be clearer, something along the lines of:

> A collection of servers that receive routed requests from the load balancer

File: source/network/load-balancers/key-terms.md
Comment (bluemorbo) [inline source/network/load-balancers/key-terms.md]: The second half of this is a little unclear, how about:

> A target is an individual server that receives a request from the load balancer

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]:
```suggestion
From the load balancer overview screen, you can work your way through the onboarding wizard. If you already know how to set up a UKFast load balancer, you can skip this wizard and work your way through the screens yourself by clicking "Skip Onboarding Wizard".
```

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]:
```suggestion
There are also optional aspects that can be configured within a load balancer, for example:
```

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]:
```suggestion
* **Source:** The source IP address of the request is hashed and then this is used to determine which target server should be used. It ensures that the IP will always end up on the same target server providing no servers are added or removed.
```

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]:
```suggestion
* **Round Robin:** Sends traffic first to server A, then to server B, then to server A and so on. This has a slightly lower overhead but can lead to hot spots developing if sticky sessions are in use.
```

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]:
```suggestion
By default, the load balancer will check the default site for each target server to see if it responds with a status code between 200 and 399. If the target server doesn't respond (or responds with a status code outside this range) the target server will be marked as unhealthy and no traffic will be sent to it until it recovers.
```

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]:
```suggestion
You can use the target group configuration screen to manage which URL to hit on a target server as either a HEAD, OPTIONS or GET request if you don't want to use the default site. For example, if you have a health check URL that performs additional checks before returning a 200 you could put the host, port and path of this URL into the form.
```

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]:
```suggestion
You can set this listener to enable HSTS for visitors, this means that the website will always need to have a valid SSL for users to access the website. If your SSL certificate is ever removed or expires, the user's browser remembers that there should be an SSL certificate and stops the user from accessing the website. **If you are going to enable this setting make sure you have an SSL certificate ready to use.**
```

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]:
```suggestion
You can also enable "Redirect to HTTPS" which similar to HSTS will redirect non-HTTPS traffic to HTTPS ensuring the user's data is encrypted between their computer and the server. However, unlike HSTS this isn't enforced and you can disable this setting if needed in the future.
```

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]: I like the warning here. I think we should take this further (though you may want a second opinion) and state that every LB will receive 1 VIP out of the box.

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]:
```suggestion
A load balancer needs at least three resources configuring before the first deployment can be made:
```

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md:42]: Just a question, is there a page we can link to so users have more detail about what fields are on this screen, and what impact they have on traffic?

I'm asking as the rest of the page has lots of lovely detail and screenshots, while this bit is a little light by comparison. Not the end of the world though, especially for a first attempt.

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]:
```suggestion
* **Least Connections:** Intelligently sends traffic to the target server that has the least connections when the request is received
```

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md:59]: Incredible. :heart_eyes:

File: source/network/load-balancers/getting-started.md
Comment (bluemorbo) [inline source/network/load-balancers/getting-started.md]: This might be more a question for someone else, but if I make a target *inactive*, what happens to the existing sticky sessions? Do they get deleted immediately or do they gracefully close off?

When we have an answer, I'd tweak the sentence accordingly.

File: source/network/load-balancers/deploying-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/deploying-changes.md]:
```suggestion
Once any change has been made to the load balancer which needs to be deployed you will see the below blue box at the top of the screen. Clicking the `deployments screen` will take you to the page to deploy these changes. If you are still using the onboarding wizard then this box won't be shown, instead, the "Deploy Now" button will activate once you've finished the required steps.
```

File: source/network/load-balancers/deploying-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/deploying-changes.md]:
```suggestion
If you've made changes since the last deployment and the configuration is valid then you will get a blue "Deploy Now" button. Clicking this will deploy the configuration you have created in MyUKFast to the load balancer so that your changes start to take effect.
```

File: source/network/load-balancers/deploying-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/deploying-changes.md]:
```suggestion
* Listeners set to redirect to HTTPS without an SSL certificate also configured
```

File: source/network/load-balancers/deploying-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/deploying-changes.md]:
```suggestion
* Listeners set to redirect to HTTPS without a bind configured on port 443
```

File: source/network/load-balancers/common-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/common-changes.md]:
```suggestion
This guide will go over some of the common changes you may want to carry out in the day-to-day administration of your load balancer server setup.
```

File: source/network/load-balancers/common-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/common-changes.md]:
```suggestion
Fill the form in with the new target server's IP address and port. If it is a UKFast server you will be able to search for and select the server IP when you click the IP field. You can also give the target server an optional name to help you remember which server it is.
```

File: source/network/load-balancers/common-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/common-changes.md]:
```suggestion
If you need to edit settings such as how often to check the health of the target server, whether to check the SSL certificate on the target server or which HTTP version to use this can be done by pressing "Show Advanced Options".
```

File: source/network/load-balancers/common-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/common-changes.md]:
```suggestion
SSL certificates have expiry dates which once reached will start to show security errors to users visiting the website. You can easily replace your load balancer's SSL certificates with a new one when they expire within MyUKFast.
```

File: source/network/load-balancers/common-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/common-changes.md]:
```suggestion
You can drain a target server to stop any traffic from going to it once the current requests have finished being dealt with. This stops a user's request from being interrupted and the user potentially getting an error.
```

File: source/network/load-balancers/common-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/common-changes.md]:
```suggestion
As traffic to your website grows, you may wish to increase its capacity by registering additional servers behind your load balancer to spread the extra requests across all available hardware.
```

File: source/network/load-balancers/common-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/common-changes.md]:
```suggestion
Once you've configured your new server and are happy for it to start receiving requests, follow the instructions below to add it to the target group your load balancer sends traffic to.
```

File: source/network/load-balancers/common-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/common-changes.md:41]: Updating an SSL en masse is going to be such a pain, we should look at a way of updating it across a whole cluster for power users in the future.

File: source/network/load-balancers/common-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/common-changes.md]: Is there any concern regarding browser caching etc? Do we need to recommend that users do a hard refresh if they're still seeing issues?

File: source/network/load-balancers/common-changes.md
Comment (bluemorbo) [inline source/network/load-balancers/common-changes.md:93]: If there's enough demand, it might be worth adding a dedicated 'Drain' secondary action button to the Edit Listener screen in the future so it's even easier to do.

File: source/network/load-balancers/faqs.md
Comment (bluemorbo) [inline source/network/load-balancers/faqs.md]:
```suggestion
This is down to the nature of the load balancer itself. Where visitors used to directly visit your server, they now visit the load balancer instead and it's the load balancer that makes requests to the backend server. As such, the only IP address you'll see in logs or analytics will likely be that of the load balancer.
```

File: source/network/load-balancers/faqs.md
Comment (bluemorbo) [inline source/network/load-balancers/faqs.md]: Is this true?

Comment (bluemorbo) [inline source/network/load-balancers/faqs.md]:
```suggestion
There's an easy fix. Whereas logs and analytics will usually derive the visitor IP from the source of the request, they need to be directed to look at the standard `X-Forwarded-For` header instead. This is injected by the load balancer automatically, so should already be present.
```

Comment (bluemorbo) [reply source/network/load-balancers/faqs.md]: I'd probably get this scrutinised by Dane before it gets merged in, just to ensure this is all still true.

**Review by bluemorbo — CHANGES_REQUESTED** [review]
Overall this is excellent, and despite there being a WALL of comments, it's mostly grammar problems that Grammarly picked up for me.

Really looking forward to this going live, so once the changes have been made, let's get someone in Tech to santiy check the code as well.

File: source/network/load-balancers/getting-started.md
Comment (rbibby) [reply source/network/load-balancers/getting-started.md:42]: Not yet, could maybe add another set of pages later for each resource with extra detail on each of the settings?

Would keep the getting stated page a bit lighter then

File: source/network/load-balancers/common-changes.md
Comment (Xiol) [reply source/network/load-balancers/common-changes.md]: Can't hurt, though in most situations I don't think it'll be a problem unless someone is actively checking that the certificate has been updated.

File: source/network/load-balancers/faqs.md
Comment (Xiol) [reply source/network/load-balancers/faqs.md]: Yeah, this is injected automatically, however there are caveats:

* Works fine with HTTP connections
* Works fine with HTTPS connections that terminate at the LB or are re-encrypted by the LB before being passed onto the backend.
* Doesn't work with HTTPS passthrough (where the LB does no decryption)
* Doesn't work with TCP mode

A workaround for HTTPS passthrough or TCP mode is to enable PROXY protocol, but this requires support from the backend.

**Comment by Xiol** [comment]
I'd like to see something in the FAQ section (possibly) about `X-Forwarded-Proto`. I'd put something in myself but I don't seem to have write access, or I'm just too used to Gitlab :)

Something like:

### Q. How do I perform HTTPS redirects from my backend?

or

### Q. I'm getting a redirect loop when redirecting to HTTPS, how can I fix this?

Answer:

In some configurations, performing a HTTP->HTTPS redirect from a server behind a loadbalancer may result in your website experiencing a redirect loop. As the HTTPS connection is terminated at the loadbalancer, the backend servers only see HTTP connections and continually try to redirect the client to a HTTPS URL.

Ideally you should perform your HTTP->HTTPS redirects directly on the loadbalancer using the option available in the Listener configuration section, however if this is not an acceptable solution, you can confirm if a client is using HTTPS connections to the loadbalancer by checking the `X-Forwarded-Proto` header. This header will be present and contain the value `https` when connections to the loadbalancer are occurring over HTTPS. You can check this header to perform conditional redirects to HTTPS on you backend servers, for example:

Apache:
```
RewriteCond %{HTTP:X-Forwarded-Proto} !https
RewriteCond %{HTTPS} off
RewriteRule ^ https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301,NE]
```

Nginx:
```
if ($http_x_forwarded_proto = "http") {
    return 301 https://$server_name$request_uri;
}
```

**Comment by rbibby** [comment]
@miff2000 any ideas what I need to do about the spellcheck failing? The only instances of `html` I can see on that page are the links themselves. I did try wrapping in `<nospell>` but didn't seem to work.

**Review by miff2000 — APPROVED** [review]
LGTM. I'm ignoring the complaint of `html` in the page

**Pull request merged**
